### PR TITLE
[cmake] sync upstream FindCUDA.cmake and use more cuda targets from FindCUDAToolkit.cmake

### PR DESCRIPTION
**Pull request by cyyever**
This work is based on PR #82695. To avoid inconsistency between CMake 3.18 and newer versions, I copy the upstream FindCUDA module to Modules_CUDA_fix. In addition, I grep and replace more references of CUDA libraries with CUDA targets. Some old checks are also removed.

### CODE REVIEW

**Comment by cyyever** [comment]
@peterbell10 Any suggestions?

File: cmake/public/cuda.cmake
Comment (peterbell10) [inline cmake/public/cuda.cmake:51]: Why are you removing the PUSH-POP here?

File: cmake/Caffe2Config.cmake.in
Comment (peterbell10) [inline cmake/Caffe2Config.cmake.in:93]: Why remove this?

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (peterbell10) [inline cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake:1]: This is a second `FindCUDAToolkit.cmake`. We should only need one.

File: cmake/Caffe2Config.cmake.in
Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:93]: It causes CMAKE_MODULE_PATH set to string “old_CMAKE_MODULE_PATH” and because now I always load our version of findcudatoolkit

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (cyyever) [reply cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake:1]: I should cover it

File: cmake/public/cuda.cmake
Comment (cyyever) [reply cmake/public/cuda.cmake:51]: I think we can safely set policy without push pop

File: cmake/Caffe2Config.cmake.in
Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:93]: Good point, there is a bug, but why are we letting our internal changes to `CMAKE_MODULE_PATH` escape into the users environment?

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (peterbell10) [reply cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake:1]: 1. The plan is to delete `Modules_CUDA_fix` so I don't think it's good to move this.
2. `FindPackageHandleStandardArgs` is a standard CMake module, so your justification doesn't make sense to me.

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (peterbell10) [inline cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake:1062]: Windows failure is real, my version of `FindCUDAToolkit.cmake` has fixes not in the upstream module, including the `EXTRA_HINTS` argument here.

```
C:\actions-runner\_work\pytorch\pytorch\torch\csrc\profiler\stubs\cuda.cpp(3): fatal error C1083: Cannot open include file: 'nvToolsExt.h': No such file or directory
```

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (cyyever) [reply cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake:1]: 1. I am agree with deleting Modules_CUDA_fix
2. But currently I have to put FindCUDAToolkit.cmake on Modules_CUDA_fix because it contains "include(${CMAKE_CURRENT_LIST_DIR}/FindPackageHandleStandardArgs.cmake)" on line 868. I am not sure why upstream module use a hard-wired path to FindPackageHandleStandardArgs.cmake, but if we just want to use it without changes, we have to also copy FindPackageHandleStandardArgs.cmake to Modules.

**Comment by cyyever** [comment]
@peterbell10 You made some changes to FindCUDAToolkit.cmake. Reverting to your variant is a better idea.

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (cyyever) [reply cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake]: OK, revert to your variant

File: cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake
Comment (cyyever) [reply cmake/Modules_CUDA_fix/upstream/FindCUDAToolkit.cmake]: OK, revert to your variant

File: cmake/Caffe2Config.cmake.in
Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:93]: That makes sense. I add them back.

**Review by peterbell10 — DISMISSED** [review]

**Comment by cyyever** [comment]
cc @malfet

**Comment by cyyever** [comment]
@malfet  May need your approval

File: cmake/Dependencies.cmake
Comment (malfet) [inline cmake/Dependencies.cmake:1857]: This needs to exclude Apple (though CUDA is dead on MacOS), as `whole-archive` is not a correct option
```suggestion
        if(UNIX and NOT APPLE)
```

**Review by malfet — APPROVED** [review]
If CI is green, then so be it, though this will likely break cross-builds for Jetson or CUDA+MacOS build.

File: cmake/public/cuda.cmake
Comment (peterbell10) [inline cmake/public/cuda.cmake:69]: Why remove this error message?

Comment (cyyever) [reply cmake/public/cuda.cmake:69]: I removed them when I tried to solve the nvtx link problem, they should be recovered.

Comment (peterbell10) [reply cmake/public/cuda.cmake:69]: If this check is failing then how is the build succeeding? That doesn't sound right at all.

Comment (cyyever) [reply cmake/public/cuda.cmake:69]: So I remove the nvtoolsext check and link nvToolsExt privately. So that it passes when including cuda.cmake and doesn't actually use nvToolsExt.

Comment (cyyever) [reply cmake/public/cuda.cmake:69]: In other words, nvtoolsext is set to NOTFOUND in cpu case but isn't used at all.

**Comment by cyyever** [comment]
> If CI is green, then so be it, though this will likely break cross-builds for Jetson or CUDA+MacOS build.

I suspect that CMake will drop CUDA support on MacOS on newer versions. But we can take a try first.

File: cmake/Caffe2Config.cmake.in
Comment (peterbell10) [inline cmake/Caffe2Config.cmake.in:91]: This is wrong, the module path is different when torch is installed.

File: cmake/public/cuda.cmake
Comment (peterbell10) [reply cmake/public/cuda.cmake:69]: This is happening because you're picking the wrong `FindCUDAToolkit.cmake`, see below comment.

File: cmake/Caffe2Config.cmake.in
Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: We always use the same module as building, that is cmake/module/findcudatoolkit.cmake.  Because cuda.cmake has list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/../Modules_CUDA_fix) in line 27

Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: I moved the code of  cmake_module_path rewriting into cuda.cmake

Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:91]: `FindCUDAToolkit.cmake` is not installed under `Module_CUDA_fix`...

Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: I also used list(PREPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/../Modules) in line 28

Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: So the modules from the two directories are searched before other places

File: cmake/public/cuda.cmake
Comment (cyyever) [reply cmake/public/cuda.cmake:69]: The version of fundcudatoolkit doesn't matter. In a cpu only test case, the cuda files will be removed after building.

File: cmake/Caffe2Config.cmake.in
Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:91]: You are wrong, the install path is shown in the build log:
```
C:\Jenkins\Miniconda3\Lib\site-packages\torch\share\cmake\Caffe2\FindCUDAToolkit.cmake
```

File: cmake/public/cuda.cmake
Comment (malfet) [reply cmake/public/cuda.cmake:69]: `cmake/public/cuda.cmake` assumes too many header dependencies which are mostly not needed to build an extension. So I'm fine with not searching for NVTX headers while building extensions. Also `nvtx` is a header only library now

File: cmake/Caffe2Config.cmake.in
Comment (malfet) [reply cmake/Caffe2Config.cmake.in:91]: @peterbell10 thank you for the thorough review. Do you know if we have any tests for `libtorch` usage in CI?  (I know there are bunch of extensions tests, but probably not libtorch ones)

But on the other hand, why wants to know anything about CUDA when linking against libtorch?

**Comment by malfet** [comment]
Let's figure out extension in libtorch story here and reflect it in PR description before landing the change.
IMO, developers using `libtorch` doesn't really need a working cuda installation, unless `#include <torch.h>` somehow depends on cuda headers (which it should not) So making cudnn and others private dependencies makes sense to me

Extensions developers is a slightly different story, so bundling the same FIndCUDA module for those makes sense, but once again, I don't think it needs to fail if say cudnn or nvtx is not found.

But if we are making those changes, we probably should say, that PR is BC breaking.

File: cmake/Caffe2Config.cmake.in
Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:91]: > Do you know if we have any tests for libtorch usage in CI? 

Yes there is `test/custom_backend` and `test/custom_operator` which use `find_package(Torch)` to build executables.

> But on the other hand, why wants to know anything about CUDA when linking against libtorch?

I believe CMake needs to have all linked targets available, even for private dependencies. The reason, quoting from the [cmake mailing list](https://cmake.org/pipermail/cmake/2016-May/063400.html):

> Static libraries do not carry information about other libraries they depend on. For this reason, when A links in B as PRIVATE and another target C links in A, CMake will still add B to the list of libraries to be linked for C because parts of B are needed by A, but A itself doesn't have that dependency encoded into it. So even though B is an internal implementation detail of A, C still needs B added to the linker command, which CMake conveniently handles for you.

Perhaps there is a way to work around this for shared libraries though.

Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: CMake needn't find private libraries again in exposed caffe2.cmake and torch.cmake in principle . Of course, the user needs to ensure the so files are in PATH.  I would like to make dependencies private as possible and remove finding code from public config files. It not only simplifies the code and avoids many unnecessary errors in rare environments such as force_cpu

Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: > You are wrong, the install path is shown in the build log:
> 
> ```
> C:\Jenkins\Miniconda3\Lib\site-packages\torch\share\cmake\Caffe2\FindCUDAToolkit.cmake
> ```

Thank you for pointing it out. May I put FindCUDAToolkit.cmake in torch\share\cmake\Caffe2\public so that it is in the same location with other modules and more importantly cuda.cmake will find it ?

Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:91]: Then you would need to add `${CMAKE_CURRENT_LIST_DIR}/public` to the module path. What's wrong with leaving it how it was?

**Comment by cyyever** [comment]
It makes sense to emit warnings when CUDA is unavailable for extensions in Caffe2.config unless they try to link libtorch_cuda, then further processing will fail.

File: cmake/Caffe2Config.cmake.in
Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:91]: > Your idea of protecting CMAKE_MODULE_PATH is great, I want to make sure it is also protected when building 

I strongly disagree. During build we want any submodules to pick up the same Find modules that we supply to avoid discrepancies between build configs. But when included as part of `find_package(Torch)` it would be incredibly rude to override `CMAKE_MODULE_PATH` so IMO these shouldn't be treated the same.

File: cmake/Dependencies.cmake
Comment (cyyever) [reply cmake/Dependencies.cmake:49]: Of course, we can reset the last commit. I just wanted to explore more possibilities to simplify dependencies before code review is continued.

File: cmake/Caffe2Config.cmake.in
Comment (cyyever) [reply cmake/Caffe2Config.cmake.in:91]: This PR makes sure that CMAKE_MODULE_PATH  is changed temporarily to prefer our variant of  FindCUDAToolkits.cmake in building pytorch and using **find_package(Torch)** from other projects. And the master version makes sure that CMAKE_MODULE_PATH  is changed temporarily by **find_package(Torch)** and _permanently_ when building pytorch. So the only difference in behavior is an additional protection in building pytorch.

Comment (peterbell10) [reply cmake/Caffe2Config.cmake.in:91]: @cyyever let me repeat:
> During build we want any submodules to pick up the same Find modules that we supply to avoid discrepancies between build configs.

i.e. we want it changed permanently when building pytorch.

**Pull request closed without merging**
